tensorflow/core/kernels/data/sampling_dataset_op.cc

```diff
@@ -0,0 +1,202 @@
+#include "tensorflow/core/kernels/data/dataset.h"
+#include "tensorflow/core/lib/random/random.h"
+
+namespace tensorflow {
+namespace data {
+namespace {
+
+// See documentation in ../../ops/dataset_ops.cc for a high-level
+// description of the following op.
+
+class SamplingDatasetOp : public UnaryDatasetOpKernel {
+ public:
+  explicit SamplingDatasetOp(OpKernelConstruction* ctx)
+      : UnaryDatasetOpKernel(ctx) {}
+
+ protected:
+  // Create a new SamplingDatasetOp::Dataset, and return it as the output.
+  void MakeDataset(OpKernelContext* ctx, DatasetBase* input,
+                   DatasetBase** output) override {
+    float rate;
+    OP_REQUIRES_OK(ctx, ParseScalarArgument<float>(ctx, "rate", &rate));
+    *output = new Dataset(ctx, rate, input);
+  }
+
+ private:
+  class Dataset : public DatasetBase {
+   public:
+    Dataset(OpKernelContext* ctx, float rate, const DatasetBase* input)
+        : DatasetBase(DatasetContext(ctx)), rate_(rate), input_(input) {
+      input_->Ref();
+    }
+
+    ~Dataset() override { input_->Unref(); }
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      if (rate_ == 0) {
+        return std::unique_ptr<IteratorBase>(new EmptyIterator(
+            {this, strings::StrCat(prefix, "::EmptySample")}));
+      } else {
+        return std::unique_ptr<IteratorBase>(
+            new Iterator({this, strings::StrCat(prefix, "::Sampling")}));
+      }
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return input_->output_dtypes();
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return input_->output_shapes();
+    }
+
+    string DebugString() const override { return "SamplingDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(SerializationContext* ctx,
+                              DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      Node* input_graph_node = nullptr;
+      TF_RETURN_IF_ERROR(b->AddInputDataset(ctx, input_, &input_graph_node));
+      Node* rate = nullptr;
+      TF_RETURN_IF_ERROR(b->AddScalar(rate_, &rate));
+      TF_RETURN_IF_ERROR(b->AddDataset(this, {input_graph_node, rate}, output));
+      return Status::OK();
+    }
+
+   private:
+    class EmptyIterator : public DatasetIterator<Dataset> {
+     public:
+      explicit EmptyIterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        *end_of_sequence = true;
+        return Status::OK();
+      }
+
+     protected:
+      std::shared_ptr<model::Node> CreateNode(
+          IteratorContext* ctx, model::Node::Args args) const override {
+        return model::MakeKnownRatioNode(std::move(args),
```

Comment:
this is incorrect, the ratio should be `rate_`.

Reply:
EmptyIterator is removed, so this line is removed with it.